pkg/diagnostics/grpc_tracing.go

```diff
@@ -98,73 +101,89 @@ func GRPCTraceUnaryServerInterceptor(appID string, spec config.TracingSpec) grpc
+			prefixedMetadata map[string]string
+			reqSpanAttr      map[string]string
+			isProxied        bool
+		)
 
 		ctx := ss.Context()
-		md, _ := metadata.FromIncomingContext(ctx)
-
-		vals := md[GRPCProxyAppIDKey]
-		if len(vals) == 0 {
-			return errors.Errorf("cannot proxy request: missing %s metadata", GRPCProxyAppIDKey)
-		}
-
-		targetID := vals[0]
-		wrapped := grpcMiddleware.WrapServerStream(ss)
-		sc, _ := SpanContextFromIncomingGRPCMetadata(ctx)
 
-		var spanKind trace.SpanStartOption
-
-		if appID == targetID {
+		// This middleware is shared by multiple services and proxied requests, which need to be handled separately
+		switch {
+		// For gRPC service invocation, this generates ServerSpan
+		case strings.HasPrefix(info.FullMethod, daprInternalPrefix):
 			spanKind = trace.WithSpanKind(trace.SpanKindServer)
-		} else {
+
+		// For gRPC API, this generates ClientSpan
+		case strings.HasPrefix(info.FullMethod, daprRuntimePrefix):
 			spanKind = trace.WithSpanKind(trace.SpanKindClient)
+
+		// For proxied requests, this generates a span depending on whether this is the server (target) or client
+		default:
+			isProxied = true
+			md, _ := metadata.FromIncomingContext(ctx)
+			vals := md.Get(GRPCProxyAppIDKey)
+			if len(vals) == 0 {
+				return fmt.Errorf("cannot proxy request: missing %s metadata", GRPCProxyAppIDKey)
+			}
+			if appID == vals[0] {
+				spanKind = trace.WithSpanKind(trace.SpanKindServer)
+			} else {
+				spanKind = trace.WithSpanKind(trace.SpanKindClient)
+			}
 		}
 
+		// Overwrite context
+		sc, _ := SpanContextFromIncomingGRPCMetadata(ctx)
 		ctx = trace.ContextWithRemoteSpanContext(ctx, sc)
-		ctx, span = tracer.Start(ctx, spanName, spanKind)
-
+		ctx, span = tracer.Start(ctx, info.FullMethod, spanKind)
+		wrapped := grpcMiddleware.WrapServerStream(ss)
 		wrapped.WrappedContext = ctx
-		err := handler(srv, wrapped)
-
-		addSpanMetadataAndUpdateStatus(ctx, span, info.FullMethod, appID, nil, true)
 
-		UpdateSpanStatusFromGRPCError(span, err)
-		span.End()
-
-		return err
-	}
-}
+		isSampled := span.SpanContext().IsSampled()
+		if isSampled {
+			// users can add dapr- prefix if they want to see the header values in span attributes.
+			prefixedMetadata = userDefinedMetadata(ctx)
+			if isProxied {
+				reqSpanAttr = map[string]string{
+					daprAPISpanNameInternal: info.FullMethod,
+				}
+			} else {
+				reqSpanAttr = spanAttributesMapFromGRPC(appID, nil, info.FullMethod)
+			}
+		}
 
-func addSpanMetadataAndUpdateStatus(ctx context.Context, span trace.Span, fullMethod, appID string, req interface{}, stream bool) {
-	var prefixedMetadata map[string]string
-	if span.SpanContext().IsSampled() {
-		// users can add dapr- prefix if they want to see the header values in span attributes.
-		prefixedMetadata = userDefinedMetadata(ctx)
-	}
+		err := handler(srv, wrapped)
 
-	if span.SpanContext().IsSampled() {
-		// Populates dapr- prefixed header first
-		AddAttributesToSpan(span, prefixedMetadata)
+		if isSampled {
```

Comment:
nit: Can this be merged with the above sample check for simplicity? It doesn't seem like the span object gets used in `handler` and if we care about avoiding extra work if the handler fails then the above block could be moved down here anyway, right?

Reply:
Uhm I'm not sure if there's a possibility the handler modifies the span. The context is added to `wrapped` so the handler could access that.

The code for unary RPCs does something similar too. I am not comfortable enough touching this sorry

Reply:
No worries, it just seems a little odd that we check it twice.